### DISCUSSION TITLE
[WS2122] Homework 5 Discussion

### DISCUSSION THREAD

**Discussion by changkun** [mimuc/gp-ws2021#54] [Discussion] [answered]
This thread is open for the discussion of homework 5.

Please also note that the deadline for the 5th homework project is 01.01.2022 23:59:59 (CET)

**Comment by MarlQ** [comment]
I'm having some problems with typescript, and I couldn't find a solution anywhere.
How am I supposed to circumvent this problem??
![image](https://user-images.githubusercontent.com/15791893/146934157-f3a805a6-e4c3-43e1-b1e0-c387c776c690.png)

**Reply by MarlQ** [reply]
Ok, I managed to "fix" it like this:
```
let edges : Edge[] = [];
for(let edge of this.edges) {
  if(edge) edges.push(edge);
}
edgeQueue.push(...edges);
```
I hate it.

**Reply by JohannesMerkt** [reply]
You can also fix it easier:
![grafik](https://user-images.githubusercontent.com/25324640/146980447-4f35fc52-7f93-4bff-9484-9f4f5e55bcec.png)

**Reply by MarlQ** [reply]
Thanks.. (Weird. I could've sworn it didn't work like that when I tried it before...)

**Reply by JohannesMerkt** [reply]
The if statement checks if the edge is null or not. Typescript is smart enough to understand that every edge within the if statement cant be null anymore. Thats why it works

**Reply by changkun** [reply]
It is a little less elegant. But I guess that's the only way we could write it. If one could find a better way to make halfedge.edges be typed as Edge[] rather than (Edge|null)[], this problem would exist :)

**Comment by JohannesMerkt** [comment]
I have now tried multiple implementations of the "vert collapse algorithm", but so far I always get stuck in a infinite loop in a subsequent vert collapse. Either in the Vertex.faces function or in the Vertex.halfedges function. This should mean that the for loop within these functions never returns to its original halfedge where it started:
![grafik](https://user-images.githubusercontent.com/25324640/146980909-36646e33-eb6b-4fca-9ed1-c8ab2eb60b80.png)
So something with the connectivity is wrong. I checked the twin and next connectivity but cant find an obvious error. now I dont know what to do. Any hints? Or has anyone managed to write a successful mesh reduction alogrithm?

**Reply by MarlQ** [reply]
I got a working edge collapse now at least. For me, it helped to put in some asserts (e.g. check when reassigning the vertex for the surrounding faces if the old vertex is actually part of the face and the new vertex isn't). 
This assert function is slightly more helpful:
```js
console.assert	= function(cond, text){
      if( cond )	return;
      console.log(text);
      throw new Error("Assertion failed!");
    };
```
Other than that, don't make the stupid mistake I did, which was to reduce `while(this.faces.length > this.faces.length-reducedFaces)` :D

**EDIT:** Actually, the edge collapse fails at an assert with larger meshes. It works correctly on a cube though (I had to delete all lines from the cube.obj except the v and f lines in order for it to work).

**Reply by JohannesMerkt** [reply]
Thank you for your hint havent thought of that

**Reply by changkun** [reply]
There are some cases that edge collapse might not work, such as broken geometry, boundary edges, edge error explosion, etc. In general, it is a wise idea to check if the edge error queue is maintained in a good way during the reduction.

**Reply by MarlQ** [reply]
![ThING](https://user-images.githubusercontent.com/15791893/147384308-ec957d54-f682-4bf6-b671-55afaa21bd2f.png)
(The picture is without re-assigning of indices)

I wanted to make the edge collapse work in all cases (skipping edges that would create non-manifold geometry), and implemented the basic procedure in the image. It works for the first two collapses for the cube, afterwards it crashes. It doesn't work for the bunny _at all_. I'm not sure how to continue from here. 

Currently, I say an edge is invalid if:
a) it is on the boundary
b) _or_ the overlapping neighbors of the two vertices of the edge to collapse are not exactly two.

**Reply by changkun** [reply]
> It works for the first two collapses for the cube, afterwards, it crashes. 

This sounds like a bug that I encountered before, but don't remember many details. I suspect that might be caused by incomplete collapse construction (missing an halfedge connection, or so).

Another reason might because of the invalid solution of the quadrics equation. If that happens, one could search an optimal position along the edge that is considered being removed.

**Reply by MarlQ** [reply]
I haven't really touched the quadrics yet, because I wanted to write an edge collapse that works first. I checked everything multiple times, and I'm not sure what halfedge connection I'd be missing...

**Reply by changkun** [reply]
Then that's probably the reason. I'd suggest to re-evaluate the entire code carefully as I remember last time I had such a similar bug took me quite sometime to realize what went wrong :)

**Reply by JohannesMerkt** [reply]
I ended up writing if condition all over the code to throw errors when something unexpected occurs. Then I came across a error that you might be struggling with too. 
![grafik](https://user-images.githubusercontent.com/25324640/147387555-12cc08f4-344a-4850-9578-22b9d649f83c.png)
It can occur that v4 and v3 are the same vert so i skip this case for now and continue with the next edge
EDIT: btw the edge im trying to collapse in the image is e12 between v1 and v2

**Reply by JohannesMerkt** [reply]
Also after i improved the sorting algorithm for the edges to be collapsed next this error occured less often up to not anymore

**Comment by MarlQ** [comment]
I'm having trouble understanding what this vector/matrix looks like:
![image](https://user-images.githubusercontent.com/15791893/147075711-1f838e07-b6ef-4c71-a988-ac4e9d4a0e0e.png)
`n_i` is a vector with 3 (or 4?) entries, but `-n_i^T  x_i` should be a number. This kind of composition is invalid, so is it supposed to be that `n_i` and the whole composed vector are transposed as well? Then `n_i` would be a normal vector with 4 (or 5) entries, which would make sense.

**Reply by changkun** [reply]
That's a good point, I don't remember if I mentioned this in the class.

The \bar{} above a 3D vector, means a homogenous coordinates, which extends 3D to 4D.

The \bar{\mathbf{n}}_i should be a column vector, therefore \mathbf{n}_i^\top \mathbf{x}_i represents a scalar, which is the 4th dimension of \bar{\mathbf{n}}_i.

The \bar{\mathbf{n}}_i is actually a column vector rather than a row vector. The correct formula should be:
![image](https://user-images.githubusercontent.com/5498964/147384349-c06c6215-7753-448e-ada6-a242ab063595.png)

**Reply by MarlQ** [reply]
For some reason, my errors are negative a lot of the time, even though I double-checked the calculation like 5 times now.
I calculate the face quadrics using n n^T , then I sum up all the face quadrics for each vert, then sum up the quadrics for the two edge verts to get Q.
Lastly, the error for v is v^T Q v 
What am I doing wrong??

**Comment by JohannesMerkt** [comment]
the three js simplifier is about 25 times faster than my QEM implementation. Is that expected?

**Reply by changkun** [reply]
I think the three.js's approach (Malex's approach) is intrinsically faster than QEM as it does not require solving a linear equation.

But as performance is not our major task here, it remains acceptable if our QEM is much slower than Malex's algorithm.

**Reply by JohannesMerkt** [reply]
Ok, I think the QEM has a better result though :)
![grafik](https://user-images.githubusercontent.com/25324640/147387311-965e4276-45e0-4720-81fa-e974bb0f1865.png)

**Reply by changkun** [reply]
That's true. Because QEM collapses the edge by a best-estimated vertex but Malex's approach only does a heuristic (curvature) one.

**Comment by changkun** [comment, marked as answer]
Close due to the end of the submission period.
